src/Definition/Doc.elm

```diff
@@ -65,11 +75,10 @@ type
     | Callout (Maybe Doc) Doc
       -- Table rows
     | Table (List (List Doc))
-      -- Folded isFolded summary details
       -- If folded, only summary is shown, otherwise
       -- summary is followed by details. Some renderers
       -- will present this as a toggle or clickable elipses
-    | Folded Bool Doc Doc
+    | Folded { id : Id Doc, isFolded : Bool, summary : Doc, details : Doc }
```

Comment:
This `Id Doc` field will be given a unique id when the doc is parsed from the server, either with a UUID library or perhaps by definition hash + and index.

@pchiusano I wonder if we should do this on the server such that other doc renderers doesn't have to add this for tree traversal and matching nodes?

Reply:
@hojberg Hmm, maybe I don't get this. The boolean seems like sufficient state for the model - it can initially be whatever is parsed from the server, and then when toggled, the boolean can be inverted.

Is the use of a separate id here is just to avoid having to do this update to the `Doc` tree? Or to maintain the association between the DOM element and the corresponding subtree of the `Doc`?

I think it will be nicer to keep all the model together in a way that it's impossible for it become inconsistent and impossible to ignore the state. Otherwise it seems like you'll have to keep adding random fields separate from the `Doc` tree, and then remembering to check those fields as appropriate in the rendering function. It might be okay, but let me give a few other approaches you can use that keeps all the state in the `Doc` tree and lets you update that as needed.

One is to still generate and use an id, and just update the `Doc` by searching for the subtree with that id to rewrite it. This could be done with a single recursive function - it's linear search, but then you're already paying that cost to render the doc. You could also attach ids to ALL doc subtrees, and then this function can be used generically for replacing or rewriting subtrees.

This latter approach can be done with cofree (which I don't think Elm can express) or with a pair of mutually recursive data types.

```Haskell
type Doc = Doc Id DocElem
type DocElem = Bold Doc | Italic Doc | ...

Doc.modify : Id -> (Doc -> Doc) -> Doc -> Doc
```

Another approach is having a `DocPath` type, and then a function:

```Haskell
Doc.modify : DocPath -> (Doc -> Doc) -> Doc -> Doc
-- Doc.at : DocPath -> Doc -> Maybe Doc
```

When building the DOM, you will know the `DocPath` at each point (just pass it as a parameter to your recursive rendering function) and you can attach appropriate handlers that will modify the correct subtree of the document. This approach avoids a linear search for the updates.

In any case, if using ids is the implementation strategy, I don't think that should be added by the backend, since it's really an implementation detail of the front-end.

Reply:
The use of an `Id` is a separate need from `DocFoldedToggles`. It helps to identify a node in the tree (I only added it to `Folded` for a lightweight solution, but could, as you suggest, be useful for all doc nodes). We can't identify by the data since the same sub doc might exist in another part of the tree. A path seems like a decent solution. I do something similar for constructing the FQNs for the sidebar. The main awkwardness with a path is in its construction to denote nestedness thats on the same node, like with `Table` for instance—would love to connect with you on options here.

`DocFoldedToggles` exists to avoid having to update the tree in an "overlay state" kind of way; leaving the original data untouched. There's pros and cons with regards to this and I don't feel super strongly. Modifying the bool directly would require traversing the tree twice (once for updating and again for rendering), but it's not we'll hit a huge performance issue with traversing twice, and having the separate state was more for convenience rather than performance.

The specific suggestion with regards to the backend, was contemplating how other consumers might need to modify the Folded bool and making it easier to traverse for them as well, calling out that this seems like a likely thing that'd need to happen there. With the path solution in mind, I guess we can forgo that :)